lib/web_ui/lib/src/engine/initialization.dart

```diff
@@ -64,15 +67,88 @@ void debugEmulateHotRestart() {
+  AssetManager? assetManager,
+}) async {
+  await initializeEngineServices(assetManager: assetManager);
+  await initializeEngineUi();
+}
+
+/// How far along the initialization process the engine is currently is.
+///
+/// The initialization process starts with [none] and proceeds in increasing
+/// `index` number until [initialized].
+enum DebugEngineInitializationState {
+  /// Initialization hasn't started yet.
+  uninitialized,
+
+  /// The engine is initializing its non-UI services.
+  initializingServices,
+
+  /// The engine has initialized its non-UI services, but hasn't started
+  /// initializing the UI.
+  initializedServices,
+
+  /// The engine started attaching UI surfaces to the web page.
+  initializingUi,
+
+  /// The engine has fully completed initialization.
+  ///
+  /// At this point the framework can start using the engine for I/O, rendering,
+  /// etc.
+  ///
+  /// This is the final state of the engine.
+  initialized,
+}
+
+/// The current initialization state of the engine.
+///
+/// See [DebugEngineInitializationState] for possible states.
+DebugEngineInitializationState get initializationState => _initializationState;
+DebugEngineInitializationState _initializationState = DebugEngineInitializationState.uninitialized;
+
+/// Resets the state back to [DebugEngineInitializationState.uninitialized].
+///
+/// This is for testing only.
+void debugResetEngineInitializationState() {
+  _initializationState = DebugEngineInitializationState.uninitialized;
+}
+
+/// Initializes non-UI engine services.
 ///
-/// This is only available on the Web, as native Flutter configures the
-/// environment in the native embedder.
-void initializeEngine() {
-  if (_engineInitialized) {
+/// Does not put any UI onto the page. It is therefore safe to call this
+/// function while the page is showing non-Flutter UI, such as a loading
+/// indicator, a splash screen, or in an add-to-app scenario where the host page
+/// is written using a different web framework.
+///
+/// See also:
+///
+///  * [initializeEngineUi], which is typically called after this function, and
+///    puts UI elements on the page.
+Future<void> initializeEngineServices({
+  AssetManager? assetManager,
+}) async {
+  if (_initializationState != DebugEngineInitializationState.uninitialized) {
+    assert(() {
+      throw StateError(
+        'Invalid engine initialization state. `initializeEngineServices` was '
+        'called, but the engine has already started initialization and is '
+        'currently in state "$_initializationState".'
+      );
+    }());
     return;
   }
+  _initializationState = DebugEngineInitializationState.initializingServices;
+
+  if (!useCanvasKit) {
+    scheduleMicrotask(() {
+      // Access [lineLookup] to force the lazy unpacking of line break data
+      // now. Removing this line won't break anything. It's just an optimization
+      // to make the unpacking happen while we are waiting for network requests.
+      lineLookup;
```

Comment:
Maybe add "converting these raw getters to the getBlah() + getter pattern so initializations are more explicit" to the cleanup task?

Reply:
Good point. Done.